lib/unparser/emitter.rb

```diff
@@ -496,5 +520,41 @@ def run(emitter, node = node())
       emitter.new(node, self).write_to_buffer
     end
 
+    # For binary or unary operators, check whether parentheses are needed
+    # around a nested operator expression.
+    #
+    # @return [Boolean]
+    #
+    # @api private
+    #
+    def needs_parentheses?(child, emitter, side)
+      # TODO: Later this can also be used for things like if/unless/until/while
+      # statement modifiers
+      # Right now it is just for nested operator expressions
+      if emitter.operator? && self.operator?
+        parent_prec = OPERATOR_PRECEDENCE[self.operator]
+        child_prec  = OPERATOR_PRECEDENCE[emitter.operator]
+        parent_intuit = INTUITIVE_PRECEDENCE[self.operator]
+        child_intuit  = INTUITIVE_PRECEDENCE[emitter.operator]
+        if parent_prec && child_prec && parent_intuit && child_intuit
+          # if omitting parens would change the meaning, we definitely need them
+          return true if child_prec < parent_prec
+          # if it is obvious to an "average" Ruby programmer that parens are not
+          # needed, then we can omit them
+          return false if child_intuit > parent_intuit
+          # in chains of identical operators, like a + b + c + d, we can omit
+          # parens if associativity allows us to do so without changing meaning
+          # + and - are a special case; a + b - c + d is OK without parens
+          return true if self.operator != emitter.operator &&
+                         (!INTUITIVE_PRECEDENCE_EQUAL.include?(self.operator) ||
+                         !INTUITIVE_PRECEDENCE_EQUAL.include?(emitter.operator))
+          associativity = OPERATOR_ASSOCIATIVITY[self.operator]
+          return false if associativity == side
```

Comment:
It's not redundant, because we fall through to the `emitter.terminated?` test.

All our tests of precedence and associativity may or may not give the answer on whether we should use parens or not. If not, we fall through to the original test: `emitter.terminated?`.